src/main/java/io/cdap/plugin/ariba/source/connector/AribaConnector.java

```diff
@@ -166,4 +185,62 @@ private JsonArray listTemplates(String pageToken, JsonArray jsonElements) throws
     }
     return jsonElements;
   }
+
+  public List<StructuredRecord> listTemplateData(String templateName) throws IOException, AribaException,
+    InterruptedException {
+    AribaServices aribaServices = new AribaServices(config);
+    AribaStructuredTransformer transformer = new AribaStructuredTransformer();
+    try {
+      accessToken = aribaServices.getAccessToken();
+    } catch (AribaException e) {
+      throw new IOException("unable to generate access token", e);
+    }
+    URL dataURL = HttpUrl.parse(config.getBaseURL()).
+      newBuilder()
+      .addPathSegments(METADATA_PATH)
+      .addPathSegments(String.format(DATA_PATH_SEGMENT, config.getSystemType()))
+      .addPathSegments(templateName)
+      .addQueryParameter(ResourceConstants.PRODUCT, ResourceConstants.ANALYTICS)
+      .addQueryParameter(ResourceConstants.REALM, config.getRealm()).build().url();
+    Request request = new Request.Builder().get().url(dataURL).addHeader(API_KEY, config.getApiKey())
+      .addHeader(AUTHORIZATION, aribaServices.getAuthenticationKey(accessToken)).build();
+    OkHttpClient enhancedOkHttpClient = new OkHttpClient();
+    Response response = enhancedOkHttpClient.newCall(request).execute();
+    List<JsonNode> nodeData = new ArrayList<>();
+    List<StructuredRecord> recordList = new ArrayList<>();
+    AribaResponseContainer aribaResponseContainer = aribaServices.tokenResponse(response);
+    Schema schema = aribaServices.buildOutputSchema(accessToken, templateName);
+    StructuredRecord structuredRecord = null;
+    try (InputStream responseStream = aribaResponseContainer.getResponseBody()) {
+      JsonNode nodeRecord = objectMapper.readTree(responseStream);
```

Comment:
can we limit the no. of records in the call or run the for loop till 1000 or sampleRequest.getLimit.
And we can skip using iterators

Reply:
If nodeRecord contain good amount of data, then we ca skip iterating that on that loop and that way we might not have to set data limit.

Reply:
Trying..